### PR TITLE
Fixed concurrency bug when etcd-druid runs more than one worker

### PR DESCRIPTION
**Pull request by georgekuruvillak**
**What this PR does / why we need it**:
This PR fixes the issue related to concurrent reconcilation of etcd resources messing up configmap, service and statefulset values.
**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```improvement operator
Fixed the issue related to concurrent reconcilation of etcd resources messing up configmap, service and statefulset values.
```

### CODE REVIEW

File: controllers/etcd_controller.go [hunk 1]
```diff
@@ -274,7 +273,7 @@ func (r *EtcdReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	}, nil
 }
 
-func (r *EtcdReconciler) reconcileServices(etcd *druidv1alpha1.Etcd) (*corev1.Service, error) {
+func (r *EtcdReconciler) reconcileServices(etcd *druidv1alpha1.Etcd, values *map[string]interface{}) (*corev1.Service, error) {
```
Comment (amshuman-kr) [inline controllers/etcd_controller.go]: Does values have to be a pointer? `maps` are already pointers.

File: controllers/etcd_controller.go [hunk 2]
```diff
@@ -355,12 +352,12 @@ func (r *EtcdReconciler) reconcileServices(etcd *druidv1alpha1.Etcd) (*corev1.Se
 	if err := controllerutil.SetControllerReference(etcd, ss, r.Scheme); err != nil {
 		return nil, err
 	}
-
+	(*values)["serviceName"] = ss.Name
```
Comment (amshuman-kr) [inline controllers/etcd_controller.go]: See [above](https://github.com/gardener/etcd-druid/pull/28/files#diff-8e5c816f8b2f1adeef36a64d81b3e0f6R276).

File: controllers/etcd_controller.go [hunk 3]
```diff
@@ -355,12 +352,12 @@ func (r *EtcdReconciler) reconcileServices(etcd *druidv1alpha1.Etcd) (*corev1.Se
...
 	return ss.DeepCopy(), err
 }
 
-func (r *EtcdReconciler) syncServiceSpec(ss *corev1.Service, etcd *druidv1alpha1.Etcd) (*corev1.Service, error) {
-	decoded, err := r.getServiceFromEtcd(etcd)
+func (r *EtcdReconciler) syncServiceSpec(ss *corev1.Service, etcd *druidv1alpha1.Etcd, values *map[string]interface{}) (*corev1.Service, error) {
```
Comment (amshuman-kr) [inline controllers/etcd_controller.go]: See [above](https://github.com/gardener/etcd-druid/pull/28/files#diff-8e5c816f8b2f1adeef36a64d81b3e0f6R276).

File: controllers/etcd_controller.go [hunk 4]
```diff
@@ -389,23 +386,28 @@ func (r *EtcdReconciler) syncServiceSpec(ss *corev1.Service, etcd *druidv1alpha1
 	return ssCopy, err
 }
 
-func (r *EtcdReconciler) getServiceFromEtcd(etcd *druidv1alpha1.Etcd) (*corev1.Service, error) {
+func (r *EtcdReconciler) getServiceFromEtcd(etcd *druidv1alpha1.Etcd, values *map[string]interface{}) (*corev1.Service, error) {
```
Comment (amshuman-kr) [inline controllers/etcd_controller.go]: See [above](https://github.com/gardener/etcd-druid/pull/28/files#diff-8e5c816f8b2f1adeef36a64d81b3e0f6R276).

File: controllers/etcd_controller.go [hunk 5]
```diff
@@ -515,23 +519,29 @@ func (r *EtcdReconciler) syncConfigMapData(cm *corev1.ConfigMap, etcd *druidv1al
...
 }
 
-func (r *EtcdReconciler) getConfigMapFromEtcd(etcd *druidv1alpha1.Etcd) (*corev1.ConfigMap, error) {
+func (r *EtcdReconciler) getConfigMapFromEtcd(etcd *druidv1alpha1.Etcd, values *map[string]interface{}) (*corev1.ConfigMap, error) {
 	var err error
 	decoded := &corev1.ConfigMap{}
 	configMapPath := getChartPathForConfigMap()
-	if _, ok := r.RenderedChart.Files()[configMapPath]; !ok {
+	chartPath := getChartPath()
+	renderedChart, err := r.ChartApplier.Render(chartPath, etcd.Name, etcd.Namespace, *values)
```
Comment (amshuman-kr) [inline controllers/etcd_controller.go]: Why render again and again? Why not do it once and pass it as a parameter to the func?

File: controllers/etcd_controller.go [hunk 6]
```diff
@@ -515,23 +519,29 @@ func (r *EtcdReconciler) syncConfigMapData(cm *corev1.ConfigMap, etcd *druidv1al
...
+	decoder := yaml.NewYAMLOrJSONDecoder(bytes.NewReader([]byte(renderedChart.Files()[configMapPath])), 1024)
 
 	if err = decoder.Decode(&decoded); err != nil {
 		return nil, err
 	}
 	return decoded, nil
 }
 
-func (r *EtcdReconciler) reconcileStatefulSet(cm *corev1.ConfigMap, svc *corev1.Service, etcd *druidv1alpha1.Etcd) (*appsv1.StatefulSet, error) {
+func (r *EtcdReconciler) reconcileStatefulSet(cm *corev1.ConfigMap, svc *corev1.Service, etcd *druidv1alpha1.Etcd, values *map[string]interface{}) (*appsv1.StatefulSet, error) {
```
Comment (amshuman-kr) [inline controllers/etcd_controller.go]: See [above](https://github.com/gardener/etcd-druid/pull/28/files#diff-8e5c816f8b2f1adeef36a64d81b3e0f6R276).

File: controllers/etcd_controller.go [hunk 7]
```diff
@@ -678,15 +687,20 @@ func (r *EtcdReconciler) syncStatefulSetSpec(ss *appsv1.StatefulSet, cm *corev1.
...
 }
 
-func (r *EtcdReconciler) getStatefulSetFromEtcd(etcd *druidv1alpha1.Etcd, cm *corev1.ConfigMap, svc *corev1.Service) (*appsv1.StatefulSet, error) {
+func (r *EtcdReconciler) getStatefulSetFromEtcd(etcd *druidv1alpha1.Etcd, cm *corev1.ConfigMap, svc *corev1.Service, values *map[string]interface{}) (*appsv1.StatefulSet, error) {
 	var err error
 	decoded := &appsv1.StatefulSet{}
 	statefulSetPath := getChartPathForStatefulSet()
-	if _, ok := r.RenderedChart.Files()[statefulSetPath]; !ok {
+	chartPath := getChartPath()
+	renderedChart, err := r.ChartApplier.Render(chartPath, etcd.Name, etcd.Namespace, *values)
```
Comment (amshuman-kr) [inline controllers/etcd_controller.go]: See [above](https://github.com/gardener/etcd-druid/pull/28/files#diff-8e5c816f8b2f1adeef36a64d81b3e0f6R527).

File: controllers/etcd_controller.go [hunk 8]
```diff
@@ -700,37 +714,17 @@ func (r *EtcdReconciler) reconcileEtcd(etcd *druidv1alpha1.Etcd) (*corev1.Servic
...
-	renderer, err = r.ChartApplier.Render(chartPath, etcd.Name, etcd.Namespace, values)
+	cm, err := r.reconcileConfigMaps(etcd, &values)
 	if err != nil {
 		return nil, nil, err
 	}
 
-	r.RenderedChart = renderer
-
-	ss, err := r.reconcileStatefulSet(cm, svc, etcd)
+	ss, err := r.reconcileStatefulSet(cm, svc, etcd, &values)
```
Comment (amshuman-kr) [inline controllers/etcd_controller.go]: See [above](https://github.com/gardener/etcd-druid/pull/28/files#diff-8e5c816f8b2f1adeef36a64d81b3e0f6R276).

File: controllers/etcd_controller.go (hunk 5 above)
Comment (georgekuruvillak) [reply controllers/etcd_controller.go]: We have to refactor extensively again. I shall optimise this in that cycle.

File: controllers/etcd_controller.go (hunk 1 above)
Comment (georgekuruvillak) [reply controllers/etcd_controller.go]: done

**Pull request merged**